cw/cw305/capture.py

```diff
@@ -152,6 +175,102 @@ def capture_aes_random(ot, ktp):
+            scope.num_segments = min(rem_num_traces, scope.num_segments_max)
+            scope.arm()
+
+            # Start batch encryption.
+            ot.target.simpleserial_write(
+                "b", scope.num_segments_actual.to_bytes(4, "little")
+            )
+
+            # Transfer traces
+            waves = scope.capture_and_transfer_waves()
+            assert waves.shape[0] == scope.num_segments
+            # Check that the ADC didn't saturate when recording this batch.
+            check_range(waves, ot.scope.adc.bits_per_sample)
+
+            # Generate plaintexts and ciphertexts to compare with the batch encryption results.
+            plaintexts = [ktp.next()[1] for _ in range(scope.num_segments_actual)]
+            ciphertexts = [cipher.encrypt(bytes(plaintexts[ii]))
+                           for ii in range(scope.num_segments_actual)]
+            # Check the last ciphertext of this batch to make sure we are in sync.
+            actual_last_ciphertext = ot.target.simpleserial_read("r", 16, ack=False)
+            expected_last_ciphertext = ciphertexts[-1]
+            assert actual_last_ciphertext == expected_last_ciphertext, (
+                f"Incorrect encryption result!\n"
+                f"actual: {actual_last_ciphertext}\n"
+                f"expected: {expected_last_ciphertext}"
+            )
+
+            # Make sure to allocate sufficient memory for the storage segment array during the
+            # first resize operation. By default, the ChipWhisperer API starts every new segment
+            # with 1 trace and then increases it on demand by 25 traces at a time. This results in
+            # frequent array resizing and decreasing capture rate.
+            # See addWave() in chipwhisperer/common/traces/_base.py.
+            if project.traces.cur_seg.tracehint < project.traces.seg_len:
+                project.traces.cur_seg.setTraceHint(project.traces.seg_len)
+            # Only keep the latest two trace storage segments enabled. By default the ChipWhisperer
+            # API keeps all segments enabled and after appending a new trace, the trace ranges are
+            # updated for all segments. This leads to a decreasing capture rate after time.
+            # See:
+            # - _updateRanges() in chipwhisperer/common/api/TraceManager.py.
+            # - https://github.com/newaetech/chipwhisperer/issues/344
+            if num_segments_storage != len(project.segments):
+                if num_segments_storage >= 2:
+                    project.traces.tm.setTraceSegmentStatus(num_segments_storage - 2, False)
+                num_segments_storage = len(project.segments)
+
+            # Add traces of this batch to the project.
+            for wave, plaintext, ciphertext in zip(waves, plaintexts, ciphertexts):
+                project.traces.append(
+                    cw.common.traces.Trace(wave, plaintext, bytearray(ciphertext), key)
+                )
```

Comment:
nit: can't remember if this was a recent change but this call is missing `dtype=np.uint16`. pointing out to avoid surprises down the road.

Reply:
When I try to add `np.uint16` as an argument to `cw.common.traces.Trace()` function, it gives an error saying "takes 5 positional arguments but 6 were given".

Reply:
The `dtype=np.uint16` argument is required for the `append()` function not the `Trace()`. Without this argument, the trace files will again be 4x bigger. This has been added very recently and it probably got forgotten when you rebased your work. 

This is a good example why we don't want to change more than one thing at once/per PR. Please read this section from the contribution guide to learn how we structure PRs and commits: https://docs.opentitan.org/doc/project/detailed_contribution_guide/#working-with-pull-requests

Reply:
Good catch @alphan by the way!

Reply:
Cool, I've tried out the PR. 100000 traces are ~235 MiB in size. So everything as expected ;-)